Generic replica connectivity

jcs47 opened the pull request:
This pull request aims to make the replica connectivity more generic. The idea is to allow other projects to re-use the connectivity while having the liberty to supply their own ordering protocol.  For example, one could want to implement another protocol that isn't MinBFT but want to use the connectivity implemented in this project.

To this end, the functions `incomingMessageHander` and `replyCollector` were replaced by interfaces `ProtocolHandler` and `ReplyCollector`, respectivly. Moreover, there are also the following new functions:

```
func GetReplica(id uint32, configer api.Configer, stack Stack, handler api.ProtocolHandler, logger *logging.Logger) (*Replica, error)
func GetClient(id uint32, n, f uint32, stack Stack, collector ReplyCollector) (Client, error)
func GetMinBFTHandler(id uint32, log messagelog.MessageLog, config api.Configer, stack Stack, logger *logging.Logger) api.ProtocolHandler
func GetLogger(id uint32, opts ...Option) *logging.Logger
func GetMinBFTCollector(f uint32) ReplyCollector
```

Commit:
core/internal: Introduce package to track view state

This internal package represents the internal state maintained by each
replica for the view-change mechanism and operations on that state.

Basically, a replica can request a view change by sending a
ReqViewChange message to its peer replicas upon expiration of an
internal timeout. To prevent sending duplicate or late ReqViewChange
messages, a replica uses RequestViewChange method to check the last
requested view change.

Upon receiving enough ReqViewChange messages, a replica will actually
stop accepting messages for the old view and start the view-change
operation by sending its ViewChange message. The ViewChange message
should include all previously generated messages with UI and get the
next UI counter value assigned to itself. Since messages are received
and handled concurrently, a replica, before generating the ViewChange
message, must ensure that all other messages are fully processed and
any corresponding generated message got UI assigned and appended to
the message log. StartViewChange method will help to synchronize this.

Upon receiving enough ViewChange messages, the primary of the new view
will complete the view change and send its NewView message. The backup
replicas will complete the view change upon receiving the NewView
message. The view state will be updated correspondingly with
FinishViewChange method.

Until this last transition, a replica cannot start processing any of
Request, Prepare or Commit messages in the new view; it has to wait
until the view change is complete. WaitAndHoldActiveView and
WaitAndHoldView methods will be used to wait until the messages can be
processed in the new view. They will also delay any view-change
operation until any already started message processing is complete.

It is possible that a replica switches to a new view even before
requesting or starting the view change itself, e.g. on receiving a
valid NewView message early from a new primary. Therefore, any late
attempt to request or start the view change should be suppressed. This
why RequestViewChange and StartViewChange additionally check against
the last finished view change.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>
Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

Commit:
core: Use viewstate package for message handling

Change the way the view number is checked when handling Request,
Prepare and Commit messages. As opposed to just retrieving the current
active view number, use the new wait and hold mechanisms provided by
viewstate package.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>
Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

Commit:
core: Start request timer in primary

Even though any replica can always assume itself to be correct, a
primary cannot safely assume its view to continue arbitrary long.
Thus, it also needs to trigger view change in case some request times
out in its view.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>
Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

Commit:
This commits aims to make the replica connectivity more generic. The idea is to allow other projects to re-use the connectivity while having the liberty to supply their own ordering protocol. Before these changes, the MinBFT protocol would always be instantiated/used, but now it can be replaced by a third party that uses this project.

To this end, the functions `incomingMessageHander` and `replyCollector` were replaced by interfaces `ProtocolHandler` and `ReplyCollector`, respectivly. Moreover, there are also $

```
func GetReplica(id uint32, configer api.Configer, stack Stack, handler api.ProtocolHandler, logger *logging.Logger) (*Replica, error)
func GetClient(id uint32, n, f uint32, stack Stack, collector ReplyCollector) (Client, error)
func GetMinBFTHandler(id uint32, log messagelog.MessageLog, config api.Configer, stack Stack, logger *logging.Logger) api.ProtocolHandler
func GetLogger(id uint32, opts ...Option) *logging.Logger
func GetMinBFTCollector(f uint32) ReplyCollector
```

Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

nhoriguchi commented on client/client.go:
```diff
@@ -64,7 +74,34 @@ func New(id uint32, n, f uint32, stack Stack) (Client, error) {
+// New creates an instance of Client given a client ID, total number
+// of replica nodes n, number of tolerated faulty replica nodes f, and
+// a stack of external interfaces.
+func GetClient(id uint32, n, f uint32, stack Stack, collector ReplyCollector) (Client, error) {
```
We already defines `api.ReplicaConnector` in `type Stack`, so I think that we can choose needed replica collector via that interface.

sergefdrv replied:
@Naoya-Horiguchi I guess the core idea here was to make `GetClient` a generic constructor for client part of BFT protocol. Then it be specialized with a specific `ReplyConnector` implementation, according to the specific BFT protocol used.

jcs47 replied:
@Naoya-Horiguchi That is a good point. Maybe it would be better to supply the collector in the stack type.
@sergefdrv , do you mean `ReplyConnector`? if, so, yes, that is the idea.

sergefdrv replied:
As far as I can see, `GetClient` is supposed to be a generic version of `New`. If we change the definition of `Stack`, we wouldn't have a default constructor for MinBFT client, i.e. `New` would become the same as `GetClient`. We probably want to keep a default constructor for MinBFT.

jcs47 replied:
Yes, that was the intention. To preserve a default constructor for MinBFT and another to supply another protocol.

sergefdrv commented:
@jcs47 First, a few general remarks. Please follow [the convention](https://git-scm.com/docs/git-commit#_discussion) to begin a commit message with a short commit title. (This reminds me of #28...) You probably wanted to submit a [_draft_ pull request](https://help.github.com/en/articles/about-pull-requests#draft-pull-requests).

nhoriguchi commented on core/replica.go:
```diff
@@ -51,6 +52,33 @@ type Replica struct {
 }
 
+// GetReplica creates a new instance of replica node
+func GetReplica(id uint32, configer api.Configer, stack Stack, handler api.ProtocolHandler, logger *logging.Logger) (*Replica, error) {
```
This also seems to me to introduce code duplication, maybe adding `api.ProtocolHandler` into `type replicaStack` is more suitable.

sergefdrv replied:
Good point. This function is almost identical to `New`. If this proposal passes, it would make sense to simply call `GetReplica` from `New`.

jcs47 replied:
Yes, adding the ProtocolHandler to replicaStack is a good design choice, but my idea was to preserve the constructor that allowed us to obtain an instance of Replica that would adopt the MinBFT protocol by default, while also providing a more generic constructor that would allow a third party to supply its own protocol. 

The key idea here is to render replica connectivity as decoupled as possible from the ordering protocol, with the goal to make that connectivity re-usable for other projects that might adopt another ordering protocol different from minbft.

nhoriguchi reviewed (commented):
Hi @jcs47, thank you for paying attention to this project. I left a few comments, but I still don't have the whole idea about how this new feature is useful. So if you revise the PR, please consider to separate the change into a series of small patches. Then, you will get more better feedback.

sergefdrv commented on client/client.go:
```diff
@@ -49,6 +50,15 @@ type Client interface {
+// the passed channel, finishes the request processing, and sends the
+// result of request execution to the passed channel.
+type ReplyCollector interface {
+	Collect(in <-chan *messages.Reply, out chan<- []byte, remover requestRemover)
```
Note that this interface depends on MinBFT-specific definition of Reply message, as well as on unexported `requestRemover` type.

jcs47 replied:
I see two options for Reply message: either change the method signature to  `Collect(in <-chan interface{}, out chan<- []byte, remover requestRemover)` (which would require additional changes throughout the code) or simply adding an extra field to the Reply message (e.g., `metadata []byte`) that would allow this type to be more generic.

Regarding the unexported `requestRemover` type, I thought about changing into to be part of the external API, but since it is created using an object from the `client/internal` package, I thought it would be best to leave it unexported.

sergefdrv replied:
> change the method signature to `Collect(in <-chan interface{}, out chan<- []byte, remover requestRemover)`

This seems to be more aligned with the whole idea of making implementation of internal connectivity generic.

> Regarding the unexported requestRemover type ... it would be best to leave it unexported.

I'm afraid external packages wouldn't be able to use the new interface because they have no access to the unexported type.

jcs47 replied:
> I'm afraid external packages wouldn't be able to use the new interface because they have no access to the unexported type.

Do you think it would be wise to make the RequestRemover external?

sergefdrv commented on client/request.go:
```diff
@@ -35,11 +35,12 @@ type requestHandler func(operation []byte) <-chan []byte
-func makeRequestHandler(clientID uint32, seq sequenceGenerator, authen api.Authenticator, buf *requestbuffer.T, f uint32) requestHandler {
+func makeRequestHandler(clientID uint32, seq sequenceGenerator, authen api.Authenticator, collector ReplyCollector, buf *requestbuffer.T, f uint32) requestHandler {
 	submitter := makeRequestSubmitter(clientID, seq, authen, buf)
-	collector := makeReplyCollector(f, buf)
```
Maybe we should rather pass `requestRemover` to `makeReplyCollector`.

jcs47 replied:
But that would force us to expose the RequestBuffer declared in `client/internal`, which is an object that is shared by requestRemover and other parts of the code that we do no want to expose. A developer that would want to supply its own ReplyCollector would have to be aware of this.

sergefdrv replied:
If we leave it as is, the user of the proposed generic connectivity would have no control over this protocol-specific functionality. A generic code should not force MinBFT-specific `requestRemover`.

jcs47 replied:
Then I wonder if we should make RequestRemover external.

sergefdrv replied:
If you envision how MinBFT-specific request remover may be generically useful.

jcs47 replied:
I'm thinking that maybe the better approach would be to enable developers to supply their own RequestHandler instead of a ReplyCollector. That way usage of a RequestRemover is at the discretion of the programmers, using whatever interface they want. It would also allow to specify how the request message is defined.

jcs47 replied:
Upon more exploration and reflection, I think there is no point in trying to make `client.New()` more generic. Developers can simply create their own implementation of the `Client` interface and re-use `ReplicaConnector`.

sergefdrv commented on client/request.go:
```diff
@@ -48,34 +49,44 @@ func makeRequestHandler(clientID uint32, seq sequenceGenerator, authen api.Authe
+	return &replyCollector{
+
+		faults: f,
+		action: collectReplies,
```
`collectReplies` could simply be a method attached to `replyCollector` type.

jcs47 replied:
Yeah, good point.

sergefdrv commented on api/api.go:
```diff
@@ -22,6 +22,16 @@ import (
+// If there is any message produced in reply, it will be send to reply
+// channel, otherwise nil channel is returned. The return value new
+// indicates that the message has not been processed before.
+type ProtocolHandler interface {
```
This package was supposed to define the API between MinBFT core and external components, but not components of the core. If we decide to expose those components, it would be better to move those to separate packages, e.g. under `core`. Then those packages would define their API.

jcs47 replied:
I am not sure I understand the idea. Wouldn't be enough to just move `ProtocolHandler` to `core`?

sergefdrv replied:
Either to `core` or maybe better to a new sub-package under `core`. That sub-package might encapsulate all related code, then.

jcs47 replied:
Wouldn't it make sense to put `ProtocolHandler` under `core` and move the code specific to minbft to either `core/minbft` or `core/default`?

sergefdrv commented on core/replica.go:
```diff
@@ -51,6 +52,33 @@ type Replica struct {
+
+		stack: stack,
+
+		log: messagelog.New(),
```
This field is used in `replica.start`, but it is not accessible by `handler`.

jcs47 replied:
It was already like this in the master repository.

sergefdrv replied:
Yes, but generic connectivity should also allow to somehow distribute messages generated by `ProtocolHandler`.

sergefdrv commented on core/replica.go:
```diff
@@ -51,6 +52,33 @@ type Replica struct {
+
+		log: messagelog.New(),
+
+		handleStream: makeMessageStreamHandler(handler, logger),
```
Basically, this would allow to reuse code from `makeMessageStreamHandler`, but that function depends on message definitions which are MinBFT-specific. Perhaps, if we make the message definition generic enough, that function would become quite trivial.

jcs47 replied:
I think the workflow implemented by the function is both generic and useful enough, maybe it suffices to adapt it to work with `interface {}` instead of `*messages.Message`?

sergefdrv replied:
I suggest you try and see if it works.

sergefdrv replied:
This might even be the most critical piece of feedback to start addressing.

jcs47 replied:
The solution I proposed seems to work. Please check commit 40ba273008f3263bd43686648006b1fe82df0fdb, which provides additional methods to the ProtocolHandler interface to serialize/deserialize messages. Don't mind the absence of documentation for the Wrap method, it is dealt with in the subsequent commit.

sergefdrv replied:
It seems to be a bit too much fuss around trying to reuse such a simple function as `makeMessageStreamHandler`. The external API tries to be generic enough to handle most of generic stuff. Maybe try to push some more generic code behind the external API? Nevertheless, if we are to go with the original proposal, it might make sense to start with some refactoring of the original code. 

Firstly, we could define a high-level abstract message type (interface) that knows how to Marshal/Unmarshal itself, as well as to give its human-readable text representation. Then this would be the job of the message implementation, not of `ProtocolHandler`. To support unmarshalling, `makeMessageStreamHandler` would receive as a parameter a mechanism to create empty messages (message factory). Empty messages would be initialized by unmarshalling the received bytes.

Secondly, we could allow for any number of reply messages, not just only one. Think of reversing the direction of peer message flow. A replica would receive from its peer just a single Hello message, then reply with a stream of messages from its message log.

jcs47 replied:
After checking out your pull request, I think an even simpler approach would be to just use a `Replica` interface. So instead of using `minbft.New`, developers that just want to re-use the connectivity could just make their own implementation of that interface. I was trying to make this in such a way that we could explicitly delimit the protocol in an interface, but maybe this is simply too much of a fuss like you speculate.

sergefdrv replied:
So what's your plan regarding this PR?

jcs47 replied:
I would suggest your PR to take priority over this one. Once it is merged with the master branch, l will pull the commits and work with what you added. If I find that none of what I proposed in ths PR is useful after the update, then this should simply be closed.

sergefdrv reviewed (changes requested):
If I understood correctly, the main idea here was to separate MinBFT-sepecific implementation from more generic code that interacts with the external interfaces. This could potentially allow to reuse those parts separately elsewhere. So far, the "generic" part does not appear really generic; and it is not quite clear yet if that would eventually reduce to something rather trivial. I was also not sure how that would agree with the scope of the project. Nevertheless, such restructuring might potentially improve overall code structure.

sergefdrv commented:
> I was also not sure how that would agree with the scope of the project.

@ynamiki @Naoya-Horiguchi What do you think?

Commit:
Augmented ProtocolHandler with Wrap/Unwrap methods, to allow different ordering protocols to use their own message definitions.

Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

Commit:
Addendum: documentation for the Wrap method

Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

Commit:
Reverted changes at client side related to allowing third-party ReplyCollectors.

Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

Commit:
Augmented Stack struct with ProtocolHandler interface.
Added Start method to ProtocolHandler.
Restructured the code implementing the minbft ProtocolHandler.
Refractored  New, GetLogger and GetMinBFTHandler to NewReplica, NewLogger and NewMinBFTHandler.

Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

Commit:
Updated test unit.

Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

Commit:
Small fix to integration test

Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

Commit:
Merge branch 'master' of https://github.com/hyperledger-labs/minbft

Signed-off-by: João Sousa <jcs@lasige.di.fc.ul.pt>

sergefdrv commented:
@jcs47 FYI, to maintain a coherent `git blame`-able history, we do not address review comments by adding more follow-up commits. I will soon prepare a proper contribution guideline (#28). The best we have so far is the discussion in #12.

ynamiki commented:
I would like to focus on the core protocol part; we have many features to be implemented for practical use in the part. And I am afraid that it is too early to generalize based on current implementation, which is under the development. Sorry for my late response.

> > I was also not sure how that would agree with the scope of the project.
> 
> @ynamiki @Naoya-Horiguchi What do you think?

sergefdrv commented:
> I would like to focus on the core protocol part; we have many features to be implemented for practical use in the part. And I am afraid that it is too early to generalize based on current implementation, which is under the development.

Should we close this PR?

ynamiki commented:
@Naoya-Horiguchi I would like to hear your comments.

nhoriguchi commented:
> @Naoya-Horiguchi I would like to hear your comments.

I'm OK to close this PR now because base code is updated (not because the original proposal is rejected). This feature is not in the roadmap, but that's not a good reason to reject it. So we can simply wait that new version of PR is submitted.

sergefdrv commented:
@jcs47 We close this PR as outdated. Please submit an updated version, if you decide to.
